packages/tools/src/tools/annotation/LivewireContourTool.ts

```diff
@@ -0,0 +1,772 @@
+      const closestHandlePoint = {
+        index: -1,
+        distSquared: Infinity,
+      };
+
+      for (let i = 0, len = controlPoints.length; i < len; i++) {
+        const controlPoint = controlPoints[i];
+        const sliceIndex = 1; // TODO: change it for volume viewports
+        const imagePoint = [controlPoint[0], controlPoint[1], sliceIndex];
+        const worldControlPoint = csUtils.transformIndexToWorld(
+          vtkImageData,
+          imagePoint
+        );
+        const canvasControlPoint = viewport.worldToCanvas(worldControlPoint);
+
+        const distSquared = math.point.distanceToPointSquared(
+          canvasPos,
+          canvasControlPoint
+        );
+
+        if (
+          distSquared <= CLICK_CLOSE_CURVE_SQR_DIST &&
+          distSquared < closestHandlePoint.distSquared
+        ) {
+          closestHandlePoint.distSquared = distSquared;
+          closestHandlePoint.index = i;
+        }
+      }
+
+      if (closestHandlePoint.index === 0) {
+        closePath = true;
+      }
+    }
+
+    this.editData.closed = this.editData.closed || closePath;
+
+    const imagePos = csUtils.transformWorldToIndex(vtkImageData, worldPos);
+    const slicePos: Types.Point2 = [imagePos[0], imagePos[1]];
+
+    this.editData.confirmedPath = this.editData.currentPath;
+
+    // Add the current cursor position as a new control point after clicking
+    this.editData.confirmedPath.addControlPoint(
+      this.editData.currentPath.getLastPoint()
+    );
+
+    // Start a new search starting at the last control point
+    this.scissors.startSearch(slicePos);
+
+    annotation.invalidated = true;
+    triggerAnnotationRenderForViewportIds(renderingEngine, viewportIdsToRender);
+
+    if (this.editData.closed) {
+      // Update the annotation because `editData` will be set to null
+      this._updateAnnotation(element, this.editData.confirmedPath);
+      this._endCallback(evt);
+    }
+
+    evt.preventDefault();
+  };
+
+  private _mouseMoveCallback = (evt: EventTypes.InteractionEventType): void => {
+    const { element, currentPoints } = evt.detail;
+    const { world: worldPos, canvas: canvasPos } = currentPoints;
+    const { viewport, renderingEngine } = getEnabledElement(element);
+    // const { annotation } = this.editData;
+    // const { data } = annotation;
+    const viewportIdsToRender = getViewportIdsWithToolToRender(
+      element,
+      this.getToolName()
+    );
+
+    this.editData.lastCanvasPoint = canvasPos;
+
+    const defaultActor = (<Types.IStackViewport>viewport).getDefaultActor();
+    const { actor } = defaultActor;
+    const vtkImageData = actor.getMapper().getInputData();
+    const dimensions = vtkImageData.getDimensions();
+    const [imgWidth, imgHeight] = dimensions;
+
+    // TODO: See what to do when working with volumes because imagePos should be
+    // the index in volume space
+    const imagePos = csUtils.transformWorldToIndex(vtkImageData, worldPos);
```

Comment:
This code was refactored after adding support for volume viewports because now there a function to convert points to/from "slice" space which does not change anything for stacks but allow us to take one slice from a volume viewport and map (0,0) and (width-1, height-1) to top-left and bottom-right corners instead of working on index space (ijk, full volume).